docs: add missing .mdx files for package-json (#2785)
flint-fyi/flint: packages/site/src/content/docs/rules/package-json/typesPresence.mdx

---
description: "Enforces that the `types` property is present."
title: "typesPresence"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="package-json" rule="typesPresence" />

This rule requires the `types` field in `package.json` to be present.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```json
{}
```

```json
{
	"name": "my-repository"
}
```

</TabItem>
<TabItem label="✅ Correct">

```json
{
	"types": "./index.d.ts"
}
```

```json
{
	"name": "my-repository",
	"types": "./index.d.ts"
}
```

</TabItem>
</Tabs>

## Options

### `ignorePrivate`

Whether the property should still be required when the package's `private` property is `true`.
Defaults to `false`.

This option can be useful if the rule is enabled on many `package.json` files and some files differ in privacy.

## When Not To Use It

If your package generates `.d.t.s` files only in locations TypeScript will automatically detect, or does not publish TypeScript declarations, you can disable this rule.

> For example, a JavaScript-only package without declaration files may not need a `types` field.

Reviewer: It's also inferred when they're properly placed by files, and in general you should use exports.

## When Not To Use It

If your package generates `.d.t.s` files only in locations TypeScript will automatically detect, or does not publish TypeScript declarations, you can disable this rule.

> Most projects generally don't need this rule because they use files with sibling declarations detected as `exports` and/or `main`.

Reviewer: Perhaps this?

```suggestion
Most projects generally don't need this rule because sibling declarations are inferred from the file structure.
```

## When Not To Use It

If your package generates `.d.t.s` files only in locations TypeScript will automatically detect, or does not publish TypeScript declarations, you can disable this rule.

> Most projects generally don't need this rule because they use files with sibling declarations detected as `exports`.

Reviewer: Those aren't mutually exclusive IIUC. I believe if you have `main: index.js`, it'll infer `types: index.d.ts` automatically, it's not just for exports.

Author: Ah nice, I'll reword.
